### PR TITLE
Rgiunti/nkunhi/fix indexed instructions

### PR DESCRIPTION
**Pull request by rgiunti**
### Problem

- Index mismatch (indexed instructions) in fft kernel (indexed instructions) main not recognized by the benchmark test which does not return a failure

- New indexed instructions test cases have been added to trigger more corner cases:
   - eew_dest != eew_indexes data are managed in multiple VRF words while indexes in just a single word and viceversa
   - for doublebw configuration: cases in which interfaces have to deal with different number of bytes

Since adding these new kinds of test cases on top of the modifications required for fixing the fft kernel issue produced errors, a proper refactor and fix of the VLSU is proposed in this PR.

### Main changes
These new or refactored signals manage the memory address generation and load/store cycle:

- **idx_gbyte**: each memory port/FU now computes the global byte position of its next index inside the index vector. It is needed for deriving  needed VRF word, the offset inside that word, and the per-port index budget. Before, these were computed by separate arithmetics (global consumed-bytes counter, MAXEW-based shifts, round-robin budget split).
- **mem_idx_word**: computes the necessary VRF word given the idx_gbyte. Memory requests are issued only when the needed word is the one that is currently fetched. Before, a port could issue with the wrong word fetched and read another port's index.
- **mem_idx_word_ok**: comparison between the needed VRF word mem_idx_word and vs2_elem_id which is the currently fetched VRF word.
- **word_index**: offset inside the fetched VRF word.
- **mem_idx_vrf_fetch_pending**: bytes of indexes that still have to be consumed by a specific port.
- **vs2_elem_id advance**: the fetched index word advances only when at least one port still has indexes to consume and no pending port needs the current word.

For doublebw configuration, the same changes have been applied to the spatz_doublebw_vlsu with some differences to remark:

- **Interface split of bytes**: the total number of bytes to be processed are splitted among the 2 interfaces. Memory-side counters, commit-side counters and address generation now use the same consistent split. This replaces messy inline expressions.
- **idx_split_bytes**: byte offset of interface 1's first index, computed correctly in both directions (eew_idx < eew_dest and eew_idx > eew_dest). The old code only right-shifted by vsew - ew, which is a negative shift when the index is wider than the data.
- **Shared-word borrow**: when the two interfaces need the same index word, interface 1 never wins arbitration. mem_idx_word_ok lets interface 1 consume indexes from interface 0's fetched word, use_upper_fu (select idx_stream half) becomes per-FU and derived from the address match instead of a static comparison.

**Other changes**

- fix for sa_gemv tests
- removal of tests **spatzBenchmarks-rtl-hp-fmatmul_M128_N128_K128** and **spatzBenchmarks-rtl-sdotp-hp-fmatmul_M128_N128_K128** from spatzbenchmark CmakeList

### CODE REVIEW

**Comment by Navaneeth-KunhiPurayil** [comment]
So this is the only test that is failing in the double bw config - 
`spatzBenchmarks-rtl-hp-sa-gemv_M128_N4096_K512` rest all seems to be fine.

**Review by Navaneeth-KunhiPurayil — DISMISSED** [review]
With the chaining fix commit the `sa-gemv` kernel can also pass.
Also the `fconv2d` bug in existing main also can pass now in questasim.

**Comment by DiyouS** [comment]
Is [PR 108](https://github.com/pulp-platform/spatz/pull/108) contained in this PR? Or we are just fixing the problems exposed by PR 108 here? If so we can close PR108 after this one is merged

**Comment by Navaneeth-KunhiPurayil** [comment]
Yeah PR 108 can be deleted

File: hw/ip/spatz/src/spatz_vlsu.sv [hunk 1]
```diff
@@ -767,6 +754,77 @@ module spatz_vlsu
     end
   end : coalesce_proc
 
+`ifndef TARGET_SYNTHESIS
```
Comment (DiyouS) [inline hw/ip/spatz/src/spatz_vlsu.sv:757]: Thank you for adding some tracing here! 
Just have one comment; I'm not sure about your ideas here. Which way of tracing do you think is better: 1. display in the simulation; 2. dump into a file?
I personally prefer 2 because in a many-core system, showing in the simulation will flush the display. But 1 seems more straightforward in a dual-core system

File: hw/ip/spatz/src/spatz_vlsu.sv [hunk 2]
```diff
@@ -841,6 +899,21 @@ module spatz_vlsu
...
+      // We calculate how many indices are expected to be fetched from the VRF for this port
+      // This signal gives more flexibility in index fetching especially if SEW and index width are different
+
+      // E.g. vl=2 SEW=16 index_width=32 and 32b config
+      // port 0 is the only port sending memory requests
+      // But index needs to be fetched corresponding to 2 ports - port 0 and port 1
+      // This signal is used to decide if we need to fetch next index or not
+      
+      max_idx_elements = (max_elements >> mem_spatz_req.vtype.vsew) << mem_spatz_req.op_mem.ew;
+
```
Comment (DiyouS) [inline hw/ip/spatz/src/spatz_vlsu.sv]: There are some trailing spaces reported by the review dog

File: hw/ip/spatz/src/spatz_vlsu.sv [hunk 3]
```diff
@@ -1249,7 +1327,7 @@ module spatz_vlsu
               end
             endcase
 
-          mem_req_svalid[port] = rob_rvalid[port] && (!mem_is_indexed || vrf_rvalid_i[1]) && !mem_spatz_req.op_mem.is_load && (commit_insn_q.vm || v0_t_read_done);
+          mem_req_svalid[port] = rob_rvalid[port] && (!mem_is_indexed || (vrf_rvalid_i[1] && mem_idx_word_ok[port])) && !mem_spatz_req.op_mem.is_load && (commit_insn_q.vm || v0_t_read_done);
```
Comment (DiyouS) [inline hw/ip/spatz/src/spatz_vlsu.sv:1330]: Do you think it might be better to add some middle paramters and break this line? e.g.
 mem_req_svalid[port] = rob_rvalid[port] &&(!mem_is_indexed || idx_ready[port]) &&& !mem_spatz_req.op_mem.is_load && xxx;
This might help with debugging

**Review by DiyouS — DISMISSED** [review]

File: hw/ip/spatz/src/spatz_vlsu.sv (hunk 1 above)
Comment (Navaneeth-KunhiPurayil) [reply hw/ip/spatz/src/spatz_vlsu.sv:757]: Both work in my opinion, in addition to large amount of text on display, I think its more clean for AI tools its just easier to read through the files and compare (especially for specific units/harts), so when we have same SW failing on one of the HW configurations, file to file comparison is easier.

File: hw/ip/spatz/src/spatz_vlsu.sv (hunk 3 above)
Comment (Navaneeth-KunhiPurayil) [reply hw/ip/spatz/src/spatz_vlsu.sv:1330]: Yes i agree, can be included in the next vlsu fixes

**Review by Navaneeth-KunhiPurayil — APPROVED** [review]
LGTM

**Review by DiyouS — APPROVED** [review]
LGTM
